x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceServiceTests.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.ccr.repository;
+
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.shard.IllegalIndexShardStateException;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.IndexShardTestCase;
+import org.junit.Before;
+
+import java.io.IOException;
+import java.util.HashSet;
+
+public class CcrRestoreSourceServiceTests extends IndexShardTestCase {
+
+    private CcrRestoreSourceService restoreSourceService;
```

Comment:
Maybe also add a test that shows that closing a shard will also close the corresponding sessions?

In order to do this, you can override

```
protected IndexShard newShard(ShardRouting routing, ShardPath shardPath, IndexMetaData indexMetaData,
                                  @Nullable CheckedFunction<IndexSettings, Store, IOException> storeProvider,
                                  @Nullable IndexSearcherWrapper indexSearcherWrapper,
                                  @Nullable EngineFactory engineFactory,
                                  Runnable globalCheckpointSyncer,
                                  IndexEventListener indexEventListener, IndexingOperationListener... listeners) throws IOException {
```
and add `restoreSourceService` as the listener.

Alternatively, you can also extend `IndexShardTestCase` so that it's possible to plugin custom `IndexEventListener`s.

Reply:
This is actually kind of tricky? That callback is called in the `IndexService`. That abstract test case does not actually use the `IndexService`.

Reply:
oh, right. Ignore my comment then.